services/automerge/pull_auto_merge.go

```diff
@@ -0,0 +1,144 @@
+		// to get the pull request for this branch.
+		// Each pull branch starts with refs/pull/ we then go from there to find the index of the pr and then
+		// use that to get the pr.
+		var prs = make([]*models.PullRequest, 0)
+		err = nil // Could be filled with an error from an earlier run
+		if strings.HasPrefix(branch, "refs/pull/") {
+
+			parts := strings.Split(branch, "/")
+
+			prIndex, err := strconv.ParseInt(parts[2], 10, 64)
+			if err != nil {
+				return err
+			}
+
+			p, err := models.GetPullRequestByIndex(repo.ID, prIndex)
+			if err != nil {
+				// If there is no pull request for this branch, we don't try to merge it.
+				if models.IsErrPullRequestNotExist(err) {
+					continue
+				}
+				return err
+			}
+			prs = append(prs, p)
+		} else {
+			prs, err = models.GetPullRequestByHeadBranch(branch, repo, models.PullRequestStatusMergeable)
+			if err != nil {
+				// If there is no pull request for this branch, we don't try to merge it.
+				if models.IsErrPullRequestNotExist(err) {
+					continue
+				}
+				return err
+			}
+		}
+
+		for _, pr := range prs {
+
+			if pr.HasMerged {
+				log.Info("PR scheduled for auto merge is already merged [ID: %d", pr.ID)
+				continue
+			}
+
+			// Check if there is a scheduled pr in the db
+			exists, scheduledPRM, err := models.GetScheduledMergeRequestByPullID(pr.ID)
+			if err != nil {
+				return err
+			}
+			if !exists {
+				log.Info("No scheduled pull request merge exists for this pr [PRID: %d]", pr.ID)
+				return nil
+			}
+
+			// Get all checks for this pr
+			// We get the latest sha commit hash again to handle the case where the check of a previous push
+			// did not succeed or was not finished yet.
+
+			if err = pr.LoadHeadRepo(); err != nil {
+				return err
+			}
+
+			headGitRepo, err := git.OpenRepository(pr.HeadRepo.RepoPath())
+			if err != nil {
+				return err
+			}
+			defer headGitRepo.Close()
+
+			headBranchExist := headGitRepo.IsBranchExist(pr.HeadBranch)
+
+			if pr.HeadRepo == nil || !headBranchExist {
+				log.Info("Head branch of auto merge pr does not exist [HeadRepoID: %d, Branch: %s, PRID: %d]", pr.HeadRepoID, pr.HeadBranch, pr.ID)
+				return nil
+			}
+
+			// Check if all checks succeeded
+			pass, err := pullservice.IsPullCommitStatusPass(pr)
+			if err != nil {
+				return err
+			}
+			if !pass {
+				log.Info("Scheduled auto merge pr has unsuccessful status checks [PRID: %d, Commit: %s]", pr.ID, sha)
+				return nil
+			}
+
+			// Merge if all checks succeeded
+			doer, err := models.GetUserByID(scheduledPRM.UserID)
+			if err != nil {
+				return err
+			}
+
+			if err = pr.LoadBaseRepo(); err != nil {
+				return err
+			}
+
+			baseGitRepo, err := git.OpenRepository(pr.BaseRepo.RepoPath())
+			if err != nil {
+				return err
+			}
+			defer baseGitRepo.Close()
+
+			if err = pullservice.Merge(pr, doer, baseGitRepo, scheduledPRM.MergeStyle, scheduledPRM.Message); err != nil {
+				return err
```

Comment:
I think it's not necessary to return, should continue to check other pr.

Reply:
maybe should run ``IsUserAllowedToMerge`` , ``pr.IsWorkInProgress`` and ``CheckPRReadyToMerge`` again before merge this pr

Reply:
But where should the error be handled instead if we're going to try merging the other prs?

Reply:
In my opinion, this error is usually due to requested rules of the PR no longer meet, such as not enough approvs, clashed with base branch, at this time will not affect other PR. Perhaps, before auto merger PR, we should check all rules of merge that including admin can ignore conditions again, and then call``pullservice.Merge``function to merge. Thanks